dev/TabView/TabView.cpp

```diff
@@ -0,0 +1,260 @@
+static constexpr double c_scrollAmount = 50.0;
+
+TabView::TabView()
+{
+    __RP_Marker_ClassById(RuntimeProfiler::ProfId_TabView);
+
+    SetDefaultStyleKey(this);
+
+    Loaded({ this, &TabView::OnLoaded });
+    SelectionChanged({ this, &TabView::OnSelectionChanged });
+    SizeChanged({ this, &TabView::OnSizeChanged });
+}
+
+void TabView::OnApplyTemplate()
+{
+    winrt::IControlProtected controlProtected{ *this };
+
+    m_tabContentPresenter.set(GetTemplateChildT<winrt::ContentPresenter>(L"TabContentPresenter", controlProtected));
+    m_scrollViewer.set(GetTemplateChildT<winrt::FxScrollViewer>(L"ScrollViewer", controlProtected));
+    if (auto scrollViewer = m_scrollViewer.get())
+    {
+        m_scrollViewerLoadedRevoker = scrollViewer.Loaded(winrt::auto_revoke, { this, &TabView::OnScrollViewerLoaded });
+    }
+}
+
+void TabView::OnPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    winrt::IDependencyProperty property = args.Property();
+
+    if (property == s_TabWidthModeProperty)
+    {
+        UpdateTabWidths();
+    }
+}
+
+winrt::AutomationPeer TabView::OnCreateAutomationPeer()
+{
+    return winrt::make<TabViewAutomationPeer>(*this);
+}
+
+void TabView::OnLoaded(const winrt::IInspectable& sender, const winrt::RoutedEventArgs& args)
+{
+    UpdateTabContent();
+}
+
+void TabView::OnScrollViewerLoaded(const winrt::IInspectable& sender, const winrt::RoutedEventArgs& args)
+{
+    if (auto scrollViewer = m_scrollViewer.get())
+    {
+        m_scrollDecreaseButton.set(SharedHelpers::FindInVisualTreeByName(scrollViewer, L"ScrollDecreaseButton").as<winrt::RepeatButton>());
+        if (auto decreaseButton = m_scrollDecreaseButton.get())
+        {
+            m_scrollDecreaseClickRevoker = decreaseButton.Click(winrt::auto_revoke, { this, &TabView::OnScrollDecreaseClick });
+        }
+
+        m_scrollIncreaseButton.set(SharedHelpers::FindInVisualTreeByName(scrollViewer, L"ScrollIncreaseButton").as<winrt::RepeatButton>());
+        if (auto increaseButton = m_scrollIncreaseButton.get())
+        {
+            m_scrollIncreaseClickRevoker = increaseButton.Click(winrt::auto_revoke, { this, &TabView::OnScrollIncreaseClick });
+        }
+
+        if (SharedHelpers::IsRS2OrHigher())
+        {
+            if (auto scrollContentPresenter = SharedHelpers::FindInVisualTreeByName(scrollViewer, L"ScrollContentPresenter").as<winrt::ScrollContentPresenter>())
+            {
+                scrollContentPresenter.TabFocusNavigation(winrt::KeyboardNavigationMode::Once);
+            }
+        }
+    }
+
+    UpdateTabWidths();
+}
+
+void TabView::OnSizeChanged(const winrt::IInspectable& sender, const winrt::SizeChangedEventArgs& args)
+{
+    UpdateTabWidths();
+}
+
+void TabView::OnItemsChanged(winrt::IInspectable const& item)
+{
+    if (auto args = item.as< winrt::IVectorChangedEventArgs>())
+    {
+        if (args.CollectionChange() == winrt::CollectionChange::ItemRemoved && Items().Size() > 0)
+        {
```

Comment:
Lots of calls to Items().Size() in this function, can you save it into a local?

Reply:
I thought it was compile out? But yes I will.

Reply:
No, it's a method call and the compiler doesn't know it doesn't have side effects (and in reality the return value colud be different each time).